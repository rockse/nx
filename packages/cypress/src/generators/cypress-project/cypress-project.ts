import {
  addDependenciesToPackageJson,
  addProjectConfiguration,
  convertNxGenerator,
  formatFiles,
  generateFiles,
  getWorkspaceLayout,
  joinPathFragments,
  logger,
  names,
  offsetFromRoot,
  ProjectConfiguration,
  readProjectConfiguration,
  stripIndents,
  toJS,
  Tree,
  updateJson,
} from '@nrwl/devkit';
import { Linter, lintProjectGenerator } from '@nrwl/linter';
import { runTasksInSerial } from '@nrwl/workspace/src/utilities/run-tasks-in-serial';
import { getRelativePathToRootTsConfig } from '@nrwl/workspace/src/utilities/typescript';

import { join } from 'path';
import { installedCypressVersion } from '../../utils/cypress-version';
import { filePathPrefix } from '../../utils/project-name';
import {
  cypressVersion,
  eslintPluginCypressVersion,
} from '../../utils/versions';
// app
import { Schema } from './schema';

export interface CypressProjectSchema extends Schema {
  projectName: string;
  projectRoot: string;
}

function createFiles(tree: Tree, options: CypressProjectSchema) {
  // if not installed or >v10 use v10 folder
  // else use v9 folder
  const cypressVersion = installedCypressVersion();
  const cypressFiles =
    cypressVersion && cypressVersion < 10 ? 'v9-and-under' : 'v10-and-after';

  generateFiles(
    tree,
    join(__dirname, './files', cypressFiles),
    options.projectRoot,
    {
      tmpl: '',
      ...options,
      project: options.project || 'Project',
      ext: options.js ? 'js' : 'ts',
      offsetFromRoot: offsetFromRoot(options.projectRoot),
      rootTsConfigPath: getRelativePathToRootTsConfig(
        tree,
        options.projectRoot
      ),
    }
  );

  if (cypressVersion && cypressVersion < 7) {
    updateJson(tree, join(options.projectRoot, 'cypress.json'), (json) => {
      json.pluginsFile = './src/plugins/index';
      return json;
    });
  } else if (cypressVersion < 10) {
    const pluginPath = join(options.projectRoot, 'src/plugins/index.js');
    if (tree.exists(pluginPath)) {
      tree.delete(pluginPath);
    }
  }

  if (options.js) {
    toJS(tree);
  }
}

function addProject(tree: Tree, options: CypressProjectSchema) {
  let e2eProjectConfig: ProjectConfiguration;

  const detectedCypressVersion = installedCypressVersion() ?? cypressVersion;

  const cypressConfig =
    detectedCypressVersion < 10 ? 'cypress.json' : 'cypress.config.ts';

  if (options.baseUrl) {
    e2eProjectConfig = {
      root: options.projectRoot,
      sourceRoot: joinPathFragments(options.projectRoot, 'src'),
      projectType: 'application',
      targets: {
        e2e: {
          executor: '@nrwl/cypress:cypress',
          options: {
            cypressConfig: joinPathFragments(
              options.projectRoot,
              cypressConfig
            ),
            baseUrl: options.baseUrl,
            testingType: 'e2e',
          },
        },
      },
      tags: [],
      implicitDependencies: options.project ? [options.project] : undefined,
    };
  } else if (options.project) {
    const project = readProjectConfiguration(tree, options.project);

    if (!project.targets) {
      logger.warn(stripIndents`
      NOTE: Project, "${options.project}", does not have any targets defined and a baseUrl was not provided. Nx will use  
      "${options.project}:serve" as the devServerTarget. But you may need to define this target within the project, "${options.project}".
      `);
    }
    const devServerTarget =
      project.targets?.serve && project.targets?.serve?.defaultConfiguration
        ? `${options.project}:serve:${project.targets.serve.defaultConfiguration}`
        : `${options.project}:serve`;
    e2eProjectConfig = {
      root: options.projectRoot,
      sourceRoot: joinPathFragments(options.projectRoot, 'src'),
      projectType: 'application',
      targets: {
        e2e: {
          executor: '@nrwl/cypress:cypress',
          options: {
            cypressConfig: joinPathFragments(
              options.projectRoot,
              cypressConfig
            ),
            devServerTarget,
            testingType: 'e2e',
          },
          configurations: {
            production: {
              devServerTarget: `${options.project}:serve:production`,
            },
          },
        },
      },
      tags: [],
      implicitDependencies: options.project ? [options.project] : undefined,
    };
  } else {
    throw new Error(`Either project or baseUrl should be specified.`);
  }

  if (detectedCypressVersion < 7) {
    e2eProjectConfig.targets.e2e.options.tsConfig = joinPathFragments(
      options.projectRoot,
      'tsconfig.json'
    );
  }
  addProjectConfiguration(
    tree,
    options.projectName,
    e2eProjectConfig,
    options.standaloneConfig
  );
}

export async function addLinter(host: Tree, options: CypressProjectSchema) {
  if (options.linter === Linter.None) {
    return () => {};
  }

  const installTask = await lintProjectGenerator(host, {
    project: options.projectName,
    linter: options.linter,
    skipFormat: true,
    tsConfigPaths: [joinPathFragments(options.projectRoot, 'tsconfig.json')],
    eslintFilePatterns: [
      `${options.projectRoot}/**/*.${options.js ? 'js' : '{js,ts}'}`,
    ],
    setParserOptionsProject: options.setParserOptionsProject,
    skipPackageJson: options.skipPackageJson,
  });

  if (!options.linter || options.linter !== Linter.EsLint) {
    return installTask;
  }

  const installTask2 = !options.skipPackageJson
    ? addDependenciesToPackageJson(
        host,
        {},
        { 'eslint-plugin-cypress': eslintPluginCypressVersion }
      )
    : () => {};

  updateJson(host, join(options.projectRoot, '.eslintrc.json'), (json) => {
    json.extends = ['plugin:cypress/recommended', ...json.extends];
    json.overrides = [
      /**
       * In order to ensure maximum efficiency when typescript-eslint generates TypeScript Programs
       * behind the scenes during lint runs, we need to make sure the project is configured to use its
       * own specific tsconfigs, and not fall back to the ones in the root of the workspace.
       */
      {
        files: ['*.ts', '*.tsx', '*.js', '*.jsx'],
        /**
         * NOTE: We no longer set parserOptions.project by default when creating new projects.
         *
         * We have observed that users rarely add rules requiring type-checking to their Nx workspaces, and therefore
         * do not actually need the capabilites which parserOptions.project provides. When specifying parserOptions.project,
         * typescript-eslint needs to create full TypeScript Programs for you. When omitting it, it can perform a simple
         * parse (and AST tranformation) of the source files it encounters during a lint run, which is much faster and much
         * less memory intensive.
         *
         * In the rare case that users attempt to add rules requiring type-checking to their setup later on (and haven't set
         * parserOptions.project), the executor will attempt to look for the particular error typescript-eslint gives you
         * and provide feedback to the user.
         */
        parserOptions: !options.setParserOptionsProject
          ? undefined
          : {
              project: `${options.projectRoot}/tsconfig.*?.json`,
            },
        /**
         * Having an empty rules object present makes it more obvious to the user where they would
         * extend things from if they needed to
         */
        rules: {},
      },
    ];

    if (installedCypressVersion() < 7) {
      /**
       * We need this override because we enabled allowJS in the tsconfig to allow for JS based Cypress tests.
       * That however leads to issues with the CommonJS Cypress plugin file.
       */
      json.overrides.push({
        files: ['src/plugins/index.js'],
        rules: {
          '@typescript-eslint/no-var-requires': 'off',
          'no-undef': 'off',
        },
      });
    }

    return json;
  });

  return runTasksInSerial(installTask, installTask2);
}

export async function cypressProjectGenerator(host: Tree, schema: Schema) {
  const options = normalizeOptions(host, schema);
  createFiles(host, options);
  addProject(host, options);
  const installTask = await addLinter(host, options);
  if (!options.skipFormat) {
    await formatFiles(host);
  }
  return installTask;
}

function normalizeOptions(host: Tree, options: Schema): CypressProjectSchema {
  const { appsDir } = getWorkspaceLayout(host);
  const projectName = filePathPrefix(
    options.directory ? `${options.directory}-${options.name}` : options.name
  );
  const projectRoot = options.directory
    ? joinPathFragments(
        appsDir,
        names(options.directory).fileName,
        options.name
      )
    : joinPathFragments(appsDir, options.name);

  options.linter = options.linter || Linter.EsLint;
  return {
    ...options,
    projectName,
    projectRoot,
  };
}

export default cypressProjectGenerator;
export const cypressProjectSchematic = convertNxGenerator(
  cypressProjectGenerator
);
