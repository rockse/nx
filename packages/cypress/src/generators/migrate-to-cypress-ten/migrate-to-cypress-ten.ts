import { assertMinimumCypressVersion } from '@nrwl/cypress/src/utils/cypress-version';
import {
  formatFiles,
  installPackagesTask,
  joinPathFragments,
  logger,
  readProjectConfiguration,
  stripIndents,
  Tree,
  updateJson,
  updateProjectConfiguration,
} from '@nrwl/devkit';
import { forEachExecutorOptions } from '@nrwl/workspace/src/utilities/executor-options-utils';
import { CypressExecutorOptions } from '../../executors/cypress/cypress.impl';
import { cypressVersion } from '../../utils/versions';
import {
  addConfigToTsConfig,
  createNewCypressConfig,
  findCypressConfigs,
  updateProjectPaths,
  writeNewConfig,
} from './conversion.util';

export async function migrateCypressProject(tree: Tree) {
  assertMinimumCypressVersion(8);

  forEachExecutorOptions<CypressExecutorOptions>(
    tree,
    '@nrwl/cypress:cypress',
    (currentValue, projectName, target, configuration) => {
      try {
        const projectConfig = readProjectConfiguration(tree, projectName);

        const { cypressConfigPathJson, cypressConfigPathTs } =
          findCypressConfigs(tree, projectConfig, target, configuration);

        // a matching cypress ts file hasn't been made yet. need to migrate.
        if (
          tree.exists(cypressConfigPathJson) &&
          !tree.exists(cypressConfigPathTs)
        ) {
          const cypressConfigs = createNewCypressConfig(
            tree,
            projectConfig,
            cypressConfigPathJson
          );

          updateProjectPaths(tree, projectConfig, cypressConfigs);
          writeNewConfig(tree, cypressConfigPathTs, cypressConfigs);
          addConfigToTsConfig(
            tree,
            projectConfig.targets?.[target]?.configurations?.tsConfig ||
              projectConfig.targets?.[target]?.options?.tsConfig ||
              joinPathFragments(projectConfig.root, 'tsconfig.json'),
            cypressConfigPathTs
          );

          tree.delete(cypressConfigPathJson);
        }
        // ts file has been made and matching json file has been removed only need to update the project config
        if (
          !tree.exists(cypressConfigPathJson) &&
          tree.exists(cypressConfigPathTs)
        ) {
          projectConfig.targets[target].options = {
            ...projectConfig.targets[target].options,
            cypressConfig: cypressConfigPathTs,
            testingType: 'e2e',
          };

          updateProjectConfiguration(tree, projectName, projectConfig);
        }
      } catch (e) {
        logger.error(stripIndents`
NX There was an error converting ${projectName}:${target}.
You can manually update the project by following the migration guide if need be.
https://nx.dev/cypress/v10-migration-guide
  `);
        throw e;
      }
    }
  );

  updateJson(tree, 'package.json', (json) => {
    json.devDependencies['cypress'] = cypressVersion;
    return json;
  });

  await formatFiles(tree);
  return () => {
    installPackagesTask(tree);
  };
}

export default migrateCypressProject;
