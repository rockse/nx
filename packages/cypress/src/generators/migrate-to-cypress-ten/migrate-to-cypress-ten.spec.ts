import {
  addProjectConfiguration,
  joinPathFragments,
  readJson,
  readProjectConfiguration,
  Tree,
  updateJson,
  updateProjectConfiguration,
} from '@nrwl/devkit';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
import { installedCypressVersion } from '../../utils/cypress-version';
import { cypressProjectGenerator } from '../cypress-project/cypress-project';
import {
  createSupportFileImport,
  updateImports,
  updateProjectPaths,
} from './conversion.util';
import { migrateCypressProject } from './migrate-to-cypress-ten';

jest.mock('../../utils/cypress-version');

describe('convertToCypressTen', () => {
  let tree: Tree;
  let mockedInstalledCypressVersion: jest.Mock<
    ReturnType<typeof installedCypressVersion>
  > = installedCypressVersion as never;

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace(2);
    mockedInstalledCypressVersion.mockReturnValue(9);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  describe('convertCypressProject', () => {
    beforeEach(async () => {
      addProjectConfiguration(tree, 'app', {
        root: 'apps/app',
        sourceRoot: 'apps/app/src',
        targets: {
          serve: {
            executor: '@nrwl/web:file-server',
            options: {},
          },
        },
      });
      mockedInstalledCypressVersion.mockReturnValue(9);

      await cypressProjectGenerator(tree, {
        name: 'app-e2e',
        skipFormat: true,
        project: 'app',
      });
    });

    it('should update project w/defaults', async () => {
      expect(tree.exists('apps/app-e2e/cypress.json')).toBeTruthy();

      await migrateCypressProject(tree);

      expect(tree.exists('apps/app-e2e/cypress.config.ts')).toBeTruthy();
      expect(
        tree.read('apps/app-e2e/cypress.config.ts', 'utf-8')
      ).toMatchSnapshot();
      expect(readJson(tree, 'apps/app-e2e/tsconfig.json').include).toEqual([
        'src/**/*.ts',
        'src/**/*.js',
        'cypress.config.ts',
      ]);
      expect(tree.exists('apps/app-e2e/src/e2e/app.cy.ts')).toBeTruthy();
      expect(tree.exists('apps/app-e2e/src/support/e2e.ts')).toBeTruthy();
    });

    it('should update project w/customized config', async () => {
      expect(tree.exists('apps/app-e2e/cypress.json')).toBeTruthy();

      updateJson(tree, 'apps/app-e2e/cypress.json', (json) => {
        json = {
          ...json,
          baseUrl: 'http://localhost:4200',
          supportFile: false,
          pluginsFile: './src/plugins/index.js',
        };

        return json;
      });

      await migrateCypressProject(tree);

      expect(tree.exists('apps/app-e2e/cypress.config.ts')).toBeTruthy();
      expect(
        tree.read('apps/app-e2e/cypress.config.ts', 'utf-8')
      ).toMatchSnapshot();
      expect(readJson(tree, 'apps/app-e2e/tsconfig.json').include).toEqual([
        'src/**/*.ts',
        'src/**/*.js',
        'cypress.config.ts',
      ]);
      expect(tree.exists('apps/app-e2e/src/e2e/app.cy.ts')).toBeTruthy();
      expect(tree.exists('apps/app-e2e/src/support/e2e.ts')).toBeTruthy();
    });

    it('should not update a non e2e project', async () => {
      await migrateCypressProject(tree);
      expect(tree.exists('apps/app/cypress.config.ts')).toBeFalsy();
      expect(tree.exists('apps/app/src/e2e/app.cy.ts')).toBeFalsy();
      expect(tree.exists('apps/app/src/support/e2e.ts')).toBeFalsy();
    });

    it('should handle custom target names', async () => {
      expect(tree.exists('apps/app-e2e/cypress.json')).toBeTruthy();
      const pc = readProjectConfiguration(tree, 'app-e2e');
      pc.targets = {
        'e2e-custom': {
          ...pc.targets['e2e'],
        },
      };
      delete pc.targets['e2e'];
      updateProjectConfiguration(tree, 'app-e2e', pc);

      await migrateCypressProject(tree);

      expect(tree.exists('apps/app-e2e/cypress.config.ts')).toBeTruthy();
      expect(
        tree.read('apps/app-e2e/cypress.config.ts', 'utf-8')
      ).toMatchSnapshot();
      expect(readJson(tree, 'apps/app-e2e/tsconfig.json').include).toEqual([
        'src/**/*.ts',
        'src/**/*.js',
        'cypress.config.ts',
      ]);
      expect(tree.exists('apps/app-e2e/src/e2e/app.cy.ts')).toBeTruthy();
      expect(tree.exists('apps/app-e2e/src/support/e2e.ts')).toBeTruthy();
    });

    it('should infer targets with --all flag', async () => {
      expect(tree.exists('apps/app-e2e/cypress.json')).toBeTruthy();
      const pc = readProjectConfiguration(tree, 'app-e2e');
      pc.targets = {
        ...pc.targets,
        'e2e-custom': {
          ...pc.targets['e2e'],
        },
      };

      updateProjectConfiguration(tree, 'app-e2e', pc);

      await migrateCypressProject(tree);

      expect(tree.exists('apps/app-e2e/cypress.config.ts')).toBeTruthy();
      expect(
        tree.read('apps/app-e2e/cypress.config.ts', 'utf-8')
      ).toMatchSnapshot();
      expect(readJson(tree, 'apps/app-e2e/tsconfig.json').include).toEqual([
        'src/**/*.ts',
        'src/**/*.js',
        'cypress.config.ts',
      ]);
      expect(tree.exists('apps/app-e2e/src/e2e/app.cy.ts')).toBeTruthy();
      expect(tree.exists('apps/app-e2e/src/support/e2e.ts')).toBeTruthy();
      expect(
        readProjectConfiguration(tree, 'app-e2e').targets
      ).toMatchSnapshot();
    });

    it('should not break when an invalid target is passed in', async () => {
      expect(tree.exists('apps/app-e2e/cypress.json')).toBeTruthy();
      const pc = readProjectConfiguration(tree, 'app-e2e');
      pc.targets = {
        ...pc.targets,
        'e2e-custom': {
          ...pc.targets['e2e'],
        },
      };

      updateProjectConfiguration(tree, 'app-e2e', pc);

      await migrateCypressProject(tree);

      expect(tree.exists('apps/app-e2e/cypress.config.ts')).toBeTruthy();
      expect(
        tree.read('apps/app-e2e/cypress.config.ts', 'utf-8')
      ).toMatchSnapshot();
      expect(readJson(tree, 'apps/app-e2e/tsconfig.json').include).toEqual([
        'src/**/*.ts',
        'src/**/*.js',
        'cypress.config.ts',
      ]);
      expect(tree.exists('apps/app-e2e/src/e2e/app.cy.ts')).toBeTruthy();
      expect(tree.exists('apps/app-e2e/src/support/e2e.ts')).toBeTruthy();
      expect(
        readProjectConfiguration(tree, 'app-e2e').targets
      ).toMatchSnapshot();
    });

    it('should handle multiple configurations', async () => {
      expect(tree.exists('apps/app-e2e/cypress.json')).toBeTruthy();
      const pc = readProjectConfiguration(tree, 'app-e2e');
      pc.targets = {
        ...pc.targets,
        e2e: {
          ...pc.targets['e2e'],
          configurations: {
            production: {
              devServerTarget: 'target:serve:production',
            },
            static: {
              baseUrl: 'http://localhost:3000',
            },
          },
        },
      };

      updateProjectConfiguration(tree, 'app-e2e', pc);

      await migrateCypressProject(tree);

      expect(tree.exists('apps/app-e2e/cypress.config.ts')).toBeTruthy();
      expect(
        tree.read('apps/app-e2e/cypress.config.ts', 'utf-8')
      ).toMatchSnapshot();
      expect(readJson(tree, 'apps/app-e2e/tsconfig.json').include).toEqual([
        'src/**/*.ts',
        'src/**/*.js',
        'cypress.config.ts',
      ]);
      expect(tree.exists('apps/app-e2e/src/e2e/app.cy.ts')).toBeTruthy();
      expect(tree.exists('apps/app-e2e/src/support/e2e.ts')).toBeTruthy();
      expect(
        readProjectConfiguration(tree, 'app-e2e').targets['e2e']
      ).toMatchSnapshot();
    });
  });

  describe('updateProjectPaths', () => {
    const cypressConfigs = {
      cypressConfigTs: {
        e2e: {
          integrationFolder: 'src/e2e',
          supportFile: 'src/support/e2e.ts',
        },
      },
      cypressConfigJson: {
        integrationFolder: 'src/integration',
        supportFile: 'src/support/index.ts',
      },
    };

    const projectConfig = {
      root: 'apps/app-e2e',
      sourceRoot: 'apps/app-e2e/src',
    };
    const filePaths = [
      'src/integration/something.spec.ts',
      'src/integration/another.spec.ts',
      'src/integration/another.spec.js',
      'src/integration/another.spec.tsx',
      'src/integration/another.spec.jsx',
      'src/integration/another.spec.cjs',
      'src/integration/another.spec.mjs',
      'src/integration/blah/another/a.spec.ts',
      'src/integration/blah/c/a.spec.ts',
      'src/support/index.ts',
      'src/plugins/index.ts',
      'src/fixtures/example.json',
    ];

    beforeEach(() => {
      for (const path of filePaths) {
        tree.write(
          joinPathFragments(projectConfig.root, path),
          String.raw`
import { getGreeting } from '../support/app.po';

import { blah } from '../support';
const eh = require('../support')

import { blah } from "../support";
const eh = require("../support")

import { blah } from '../../support';
const eh = require('../../support')

import { blah } from "../../support";
const eh = require("../../support")
`
        );
      }
    });

    it('should rename old support file to e2e.ts', () => {
      const newSupportFile = joinPathFragments(
        projectConfig.root,
        cypressConfigs.cypressConfigTs.e2e.supportFile
      );
      const oldSupportFile = joinPathFragments(
        projectConfig.root,
        cypressConfigs.cypressConfigJson.supportFile
      );
      updateProjectPaths(tree, projectConfig, cypressConfigs);

      expect(tree.exists(newSupportFile)).toEqual(true);
      expect(tree.exists(oldSupportFile)).toEqual(false);
    });

    it('should rename files .spec. to .cy.', () => {
      const specs = tree
        .children(projectConfig.root)
        .filter((path) => path.endsWith('.spec.ts'));

      updateProjectPaths(tree, projectConfig, cypressConfigs);
      const actualSpecs = tree
        .children(projectConfig.root)
        .filter((path) => path.endsWith('.spec.ts'));
      const actualCy = tree
        .children(projectConfig.root)
        .filter((path) => path.endsWith('.cy.ts'));

      expect(actualSpecs.length).toEqual(0);
      expect(actualCy.length).toEqual(specs.length);
    });

    it('should move files to the new integration folder (e2e/)', () => {
      const newIntegrationFolder = joinPathFragments(
        projectConfig.root,
        cypressConfigs.cypressConfigTs.e2e.integrationFolder
      );
      const oldIntegrationFolder = joinPathFragments(
        projectConfig.root,
        cypressConfigs.cypressConfigJson.integrationFolder
      );
      const oldIntegrationFolderContents = tree.children(oldIntegrationFolder);

      updateProjectPaths(tree, projectConfig, cypressConfigs);

      const newIntegrationFolderContents = tree.children(newIntegrationFolder);

      expect(tree.exists(oldIntegrationFolder)).toEqual(false);
      expect(newIntegrationFolderContents.length).toEqual(
        oldIntegrationFolderContents.length
      );
      expect(tree.exists('apps/app-e2e/src/fixtures/example.json')).toEqual(
        true
      );
    });
  });

  describe('updateImports', () => {
    const filePath = 'apps/app-e2e/src/e2e/sometest.cy.ts';
    const fileContents = String.raw`
import { getGreeting } from '../support/app.po';

import { blah } from '../support';
const eh = require('../support')

import { blah } from "../support";
const eh = require("../support")

describe('a', () => {
  beforeEach(() => {
    cy.visit('/')
    blah()
    eh()
  });

  it('should display welcome message', () => {
    // Custom command example, see \`../support/commands.ts\` file
    cy.login('my-email@something.com', 'myPassword');

    // Function helper example, see \`../support/app.po.ts\` file
    getGreeting().contains('Welcome a');
  });
});

`;

    beforeEach(() => {
      tree.write(filePath, fileContents);
    });

    it('should update imports', () => {
      updateImports(tree, filePath, 'support', 'support/e2e');
      const actual = tree.read(filePath, 'utf-8');

      expect(actual).toMatchSnapshot();
    });
  });

  describe('Support File Imports', () => {
    const newImport = 'apps/app-e2e/src/support/e2e.ts';

    it('should update imports w/defaults', () => {
      const oldImport = 'apps/app-e2e/src/support/index.ts';

      const actual = createSupportFileImport(
        oldImport,
        newImport,
        'apps/app-e2e/src'
      );
      expect(actual).toEqual({
        oldImportPathLeaf: 'support',
        newImportPathLeaf: 'support/e2e',
      });
    });

    it('should handle custom support file location', () => {
      const oldImport = 'apps/app-e2e/src/support/blah.ts';

      const actual = createSupportFileImport(
        oldImport,
        newImport,
        'apps/app-e2e/src'
      );
      expect(actual).toEqual({
        oldImportPathLeaf: 'support/blah',
        newImportPathLeaf: 'support/e2e',
      });
    });

    it('should handle nested custom support location', () => {
      const oldImport = 'apps/app-e2e/src/support/blah/abc.ts';

      const actual = createSupportFileImport(
        oldImport,
        newImport,
        'apps/app-e2e/src'
      );
      expect(actual).toEqual({
        oldImportPathLeaf: 'support/blah/abc',
        newImportPathLeaf: 'support/e2e',
      });
    });

    it('should handle nested custom support location w/index.ts', () => {
      const oldImport = 'apps/app-e2e/src/support/something/neat/index.ts';

      const actual = createSupportFileImport(
        oldImport,
        newImport,
        'apps/app-e2e/src'
      );
      expect(actual).toEqual({
        oldImportPathLeaf: 'support/something/neat',
        newImportPathLeaf: 'support/e2e',
      });
    });
  });
});
