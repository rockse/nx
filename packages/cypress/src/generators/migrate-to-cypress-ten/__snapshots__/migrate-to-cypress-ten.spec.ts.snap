// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertToCypressTen convertCypressProject should handle custom target names 1`] = `
"import { defineConfig } from 'cypress'
import { nxE2EPreset } from '@nrwl/cypress/plugins/cypress-preset';


const cypressJsonConfig = {
\\"fileServerFolder\\": \\".\\",
\\"fixturesFolder\\": \\"./src/fixtures\\",
\\"video\\": true,
\\"videosFolder\\": \\"../../dist/cypress/apps/app-e2e/videos\\",
\\"screenshotsFolder\\": \\"../../dist/cypress/apps/app-e2e/screenshots\\",
\\"chromeWebSecurity\\": false,
\\"specPattern\\": \\"src/e2e/**/*.cy.{js,jsx,ts,tsx}\\",
\\"supportFile\\": \\"src/support/e2e.ts\\"
}
export default defineConfig({
e2e: {
...nxE2EPreset(__dirname),
...cypressJsonConfig,

}
})"
`;

exports[`convertToCypressTen convertCypressProject should handle multiple configurations 1`] = `
"import { defineConfig } from 'cypress'
import { nxE2EPreset } from '@nrwl/cypress/plugins/cypress-preset';


const cypressJsonConfig = {
\\"fileServerFolder\\": \\".\\",
\\"fixturesFolder\\": \\"./src/fixtures\\",
\\"video\\": true,
\\"videosFolder\\": \\"../../dist/cypress/apps/app-e2e/videos\\",
\\"screenshotsFolder\\": \\"../../dist/cypress/apps/app-e2e/screenshots\\",
\\"chromeWebSecurity\\": false,
\\"specPattern\\": \\"src/e2e/**/*.cy.{js,jsx,ts,tsx}\\",
\\"supportFile\\": \\"src/support/e2e.ts\\"
}
export default defineConfig({
e2e: {
...nxE2EPreset(__dirname),
...cypressJsonConfig,

}
})"
`;

exports[`convertToCypressTen convertCypressProject should handle multiple configurations 2`] = `
Object {
  "configurations": Object {
    "production": Object {
      "devServerTarget": "target:serve:production",
    },
    "static": Object {
      "baseUrl": "http://localhost:3000",
    },
  },
  "executor": "@nrwl/cypress:cypress",
  "options": Object {
    "cypressConfig": "apps/app-e2e/cypress.config.ts",
    "devServerTarget": "app:serve",
    "testingType": "e2e",
  },
}
`;

exports[`convertToCypressTen convertCypressProject should infer targets with --all flag 1`] = `
"import { defineConfig } from 'cypress'
import { nxE2EPreset } from '@nrwl/cypress/plugins/cypress-preset';


const cypressJsonConfig = {
\\"fileServerFolder\\": \\".\\",
\\"fixturesFolder\\": \\"./src/fixtures\\",
\\"video\\": true,
\\"videosFolder\\": \\"../../dist/cypress/apps/app-e2e/videos\\",
\\"screenshotsFolder\\": \\"../../dist/cypress/apps/app-e2e/screenshots\\",
\\"chromeWebSecurity\\": false,
\\"specPattern\\": \\"src/e2e/**/*.cy.{js,jsx,ts,tsx}\\",
\\"supportFile\\": \\"src/support/e2e.ts\\"
}
export default defineConfig({
e2e: {
...nxE2EPreset(__dirname),
...cypressJsonConfig,

}
})"
`;

exports[`convertToCypressTen convertCypressProject should infer targets with --all flag 2`] = `
Object {
  "e2e": Object {
    "configurations": Object {
      "production": Object {
        "devServerTarget": "app:serve:production",
      },
    },
    "executor": "@nrwl/cypress:cypress",
    "options": Object {
      "cypressConfig": "apps/app-e2e/cypress.config.ts",
      "devServerTarget": "app:serve",
      "testingType": "e2e",
    },
  },
  "e2e-custom": Object {
    "configurations": Object {
      "production": Object {
        "devServerTarget": "app:serve:production",
      },
    },
    "executor": "@nrwl/cypress:cypress",
    "options": Object {
      "cypressConfig": "apps/app-e2e/cypress.config.ts",
      "devServerTarget": "app:serve",
      "testingType": "e2e",
    },
  },
  "lint": Object {
    "executor": "@nrwl/linter:eslint",
    "options": Object {
      "lintFilePatterns": Array [
        "apps/app-e2e/**/*.{js,ts}",
      ],
    },
    "outputs": Array [
      "{options.outputFile}",
    ],
  },
}
`;

exports[`convertToCypressTen convertCypressProject should not break when an invalid target is passed in 1`] = `
"import { defineConfig } from 'cypress'
import { nxE2EPreset } from '@nrwl/cypress/plugins/cypress-preset';


const cypressJsonConfig = {
\\"fileServerFolder\\": \\".\\",
\\"fixturesFolder\\": \\"./src/fixtures\\",
\\"video\\": true,
\\"videosFolder\\": \\"../../dist/cypress/apps/app-e2e/videos\\",
\\"screenshotsFolder\\": \\"../../dist/cypress/apps/app-e2e/screenshots\\",
\\"chromeWebSecurity\\": false,
\\"specPattern\\": \\"src/e2e/**/*.cy.{js,jsx,ts,tsx}\\",
\\"supportFile\\": \\"src/support/e2e.ts\\"
}
export default defineConfig({
e2e: {
...nxE2EPreset(__dirname),
...cypressJsonConfig,

}
})"
`;

exports[`convertToCypressTen convertCypressProject should not break when an invalid target is passed in 2`] = `
Object {
  "e2e": Object {
    "configurations": Object {
      "production": Object {
        "devServerTarget": "app:serve:production",
      },
    },
    "executor": "@nrwl/cypress:cypress",
    "options": Object {
      "cypressConfig": "apps/app-e2e/cypress.config.ts",
      "devServerTarget": "app:serve",
      "testingType": "e2e",
    },
  },
  "e2e-custom": Object {
    "configurations": Object {
      "production": Object {
        "devServerTarget": "app:serve:production",
      },
    },
    "executor": "@nrwl/cypress:cypress",
    "options": Object {
      "cypressConfig": "apps/app-e2e/cypress.config.ts",
      "devServerTarget": "app:serve",
      "testingType": "e2e",
    },
  },
  "lint": Object {
    "executor": "@nrwl/linter:eslint",
    "options": Object {
      "lintFilePatterns": Array [
        "apps/app-e2e/**/*.{js,ts}",
      ],
    },
    "outputs": Array [
      "{options.outputFile}",
    ],
  },
}
`;

exports[`convertToCypressTen convertCypressProject should update project w/customized config 1`] = `
"import { defineConfig } from 'cypress'
import { nxE2EPreset } from '@nrwl/cypress/plugins/cypress-preset';
import setupNodeEvents from './src/plugins/index.js';

const cypressJsonConfig = {
\\"fileServerFolder\\": \\".\\",
\\"fixturesFolder\\": \\"./src/fixtures\\",
\\"video\\": true,
\\"videosFolder\\": \\"../../dist/cypress/apps/app-e2e/videos\\",
\\"screenshotsFolder\\": \\"../../dist/cypress/apps/app-e2e/screenshots\\",
\\"chromeWebSecurity\\": false,
\\"baseUrl\\": \\"http://localhost:4200\\",
\\"specPattern\\": \\"src/e2e/**/*.cy.{js,jsx,ts,tsx}\\",
\\"supportFile\\": false
}
export default defineConfig({
e2e: {
...nxE2EPreset(__dirname),
...cypressJsonConfig,
setupNodeEvents
}
})"
`;

exports[`convertToCypressTen convertCypressProject should update project w/defaults 1`] = `
"import { defineConfig } from 'cypress'
import { nxE2EPreset } from '@nrwl/cypress/plugins/cypress-preset';


const cypressJsonConfig = {
\\"fileServerFolder\\": \\".\\",
\\"fixturesFolder\\": \\"./src/fixtures\\",
\\"video\\": true,
\\"videosFolder\\": \\"../../dist/cypress/apps/app-e2e/videos\\",
\\"screenshotsFolder\\": \\"../../dist/cypress/apps/app-e2e/screenshots\\",
\\"chromeWebSecurity\\": false,
\\"specPattern\\": \\"src/e2e/**/*.cy.{js,jsx,ts,tsx}\\",
\\"supportFile\\": \\"src/support/e2e.ts\\"
}
export default defineConfig({
e2e: {
...nxE2EPreset(__dirname),
...cypressJsonConfig,

}
})"
`;

exports[`convertToCypressTen updateImports should update imports 1`] = `
"
import { getGreeting } from '../support/app.po';

import { blah } from '../support/e2e';
const eh = require('../support/e2e')

import { blah } from '../support/e2e';
const eh = require('../support/e2e')

describe('a', () => {
  beforeEach(() => {
    cy.visit('/')
    blah()
    eh()
  });

  it('should display welcome message', () => {
    // Custom command example, see \\\\\`../support/commands.ts\\\\\` file
    cy.login('my-email@something.com', 'myPassword');

    // Function helper example, see \\\\\`../support/app.po.ts\\\\\` file
    getGreeting().contains('Welcome a');
  });
});

"
`;
